ApeWorX/ape
# feat: ipc support in geth provider [APE-827]

### What I did

fixes: #1379 
Fixes: APE-815

* Attempts to use `ipc` when missing http apis (uses `web3.provider.AutoProvider` to achieve this multi-provider support).
* Adds support for `ipc_path` and `data_dir` in the `ape-config.yaml`. This helps connect to mainnet. I had to use a different data dir for my mainnet stuff. So now I can tell ape to use that data dir which is needed for the ipc stuff to work.

### How I did it

flailing around, trying to figure out what is needed to resolve and advise the issue 

### How to verify it

Start geth like this:

```shell
geth --syncmode light --networkid 1 --http --http.api "eth,debug,web3" --mainnet --datadir ~/geth_data_mainnet
```

**NOTE**: The `web3` allows HTTP to work, but since you are running geth yourself, it also allows IPC to work, and IPC is faster so that is what `Autoprovider` ends up using!

Use `ape-config.yaml` like this:

```yaml
geth:
  data_dir: ~/geth_data_mainnet
```

Start ape console:

```shell
ape console --network 'ethereum:mainnet:geth'
```

Verify Provider is IPC:

```
In [1]: provider.web3.provider._active_provider
Out[1]: <web3.providers.ipc.IPCProvider at 0x122d41610>
```

### Checklist
